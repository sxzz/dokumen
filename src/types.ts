export interface TypeRef {
  start: number
  end: number
  file: string
  text?: string
}

export interface ResolvedType {
  refs?: TypeRef[]
  unionType?: ResolvedType[]
  text: string
}

export interface Prop {
  name: string
  type: ResolvedType
  default: string
  description: string
  required: boolean
}

export interface Emit {
  name: string
  signatures: { name: string; type: ResolvedType }[][]
  description: string
}
