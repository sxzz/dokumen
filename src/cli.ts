import { writeFile } from 'fs/promises'
import path from 'path'
import { generate } from '.'

const rootPath = path.join(__dirname, '..')
const tsConfigFilePath = path.join(rootPath, 'tsconfig.json')

const results = generate(process.argv.slice(2), { rootPath, tsConfigFilePath })

for (const { result, filePath } of results) {
  const jsonFile = filePath.replace(/\.(ts|vue)$/, '.json')
  writeFile(jsonFile, JSON.stringify(result, undefined, 2))
}
