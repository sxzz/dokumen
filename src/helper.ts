export const helperCode = `
import type { ExtractPropTypes } from 'vue'
export type ResolveProp<T> = ExtractPropTypes<{
  key: { type: T; required: true }
}>['key']
export type ResolvePropType<T> = ResolveProp<T> extends { type: infer V }
  ? V
  : ResolveProp<T>
`
