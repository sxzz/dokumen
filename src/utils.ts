import { Node } from 'ts-morph'
import type { Symbol } from 'ts-morph'

export const getDescription = (
  symbol: Symbol,
  name: string,
  filter: Record<string, string> = {}
) =>
  symbol
    .getJsDocTags()
    .filter((tag) => {
      if (tag.getName() !== name) return false

      if (filter) {
        const text = tag.getText()
        return Object.entries(filter).every(([kind, value]) =>
          text.some((text) => text.kind === kind && text.text === value)
        )
      }

      return true
    })
    .map((tag) => tag.getText().find((text) => text.kind === 'text')?.text)
    .join('\n')

export const isTsLib = (filePath: string) =>
  filePath.includes('node_modules/typescript/lib')

// export const resolveType = (symbol: Symbol) => {}

export const findInitializer = (symbol: Symbol | undefined) => {
  if (!symbol) return undefined
  const decl = symbol.getValueDeclaration()
  if (Node.isInitializerExpressionGetable(decl)) {
    return decl.getInitializer()
  }
  return undefined
}
