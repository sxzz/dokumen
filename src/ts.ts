import path from 'path'
import { Node, SyntaxKind, TypeFormatFlags } from 'ts-morph'
import { findInitializer, getDescription, isTsLib } from './utils'
import { helperCode } from './helper'
import type { Options } from '.'
import type { Emit, Prop, ResolvedType, TypeRef } from './types'
import type { CallExpression, SourceFile, Type } from 'ts-morph'

export interface Result {
  componentName?: string
  props?: Prop[]
  emits?: Emit[]
}

export const parseTS = (
  sourceFile: SourceFile,
  { rootPath }: Options
): Result => {
  sourceFile.addStatements(helperCode)

  return processOptionsAPI()

  function processOptionsAPI(): Result {
    let callExpr: CallExpression<ts.CallExpression> | undefined

    function getComponentOptions() {
      let expr: Node = assignment!.getExpression()

      // variable reference
      if (expr.asKind(SyntaxKind.Identifier)) {
        const ref = findInitializer(expr.getSymbol())!
        if (!ref) throw new Error('Option API: cannot find reference')

        expr = ref
      }

      if (expr.getKind() === SyntaxKind.CallExpression) {
        callExpr = expr.asKindOrThrow(SyntaxKind.CallExpression)
        const arg = callExpr.getArguments()[0]
        if (!arg) throw new Error('Option API: cannot find call first argument')
        expr = arg
      }

      if (expr.getKind() !== SyntaxKind.ObjectLiteralExpression) {
        throw new Error('Option API: cannot find property definition')
      }

      return expr.asKindOrThrow(SyntaxKind.ObjectLiteralExpression).getType()
    }

    function getComponentName() {
      const nameDecl = componentOptions
        .getProperty('name')
        ?.getValueDeclaration()
      const nameType = nameDecl?.getType()
      const init = findInitializer(nameDecl?.getSymbol())
      if (init && Node.isStringLiteral(init)) {
        return init.getLiteralValue()
      } else if (nameType?.isStringLiteral()) {
        return nameType.getLiteralValue() as string
      }
    }

    function isDefineComponentCall() {
      return (
        callExpr?.getType()?.getAliasSymbol()?.getName() === 'DefineComponent'
      )
    }

    function analyzePropsType(propsType: Type<ts.Type>): Prop[] {
      if (!propsType.isObject()) return []

      return propsType
        .getProperties()
        .map((prop): Prop | undefined => {
          const name = prop.getName()
          const description = getDescription(prop, 'description')

          const initializer = findInitializer(prop)?.getType()
          if (!initializer) return undefined

          const typeInitializer = findInitializer(
            initializer.getProperty('type')
          )?.getType()
          let type: ResolvedType
          if (!typeInitializer) {
            type = { text: 'any' }
          } else if (
            typeInitializer.getAliasSymbol()?.getName() === 'PropType'
          ) {
            type = resolveType(typeInitializer.getAliasTypeArguments()[0])
          } else {
            type = resolvePropRawType(sourceFile, typeInitializer)
          }

          const defaultValue =
            findInitializer(initializer.getProperty('default'))
              ?.getType()
              .getText() ?? ''
          const required =
            findInitializer(initializer.getProperty('required'))
              ?.getType()
              .getText() === 'true' ?? false

          return {
            name,
            type,
            default: defaultValue,
            description,
            required,
          }
        })
        .filter((prop): prop is Prop => !!prop)
    }

    function analyzeEmitsType(emitsType: Type<ts.Type>): Emit[] {
      if (!emitsType.isObject()) return []
      return emitsType.getProperties().map((emit): Emit => {
        const name = emit.getName()
        const description = getDescription(emit, 'description')
        const _type = emit.getValueDeclarationOrThrow().getType()
        const types = _type.isUnion() ? _type.getUnionTypes() : [_type]

        const signatures: { name: string; type: ResolvedType }[][] = []
        types.forEach((type) => {
          signatures.push(
            ...type.getCallSignatures().map((sign) => {
              const args = sign.getParameters().map((param) => {
                const name = param.getName()
                const type = resolveType(
                  param.getValueDeclarationOrThrow().getType()
                )
                const description = getDescription(param, 'param', {
                  parameterName: name,
                })
                return {
                  name,
                  type,
                  description,
                }
              })
              return args
            })
          )
        })

        return {
          name,
          signatures,
          description,
        }
      })
    }

    const assignment = sourceFile.getExportAssignment(() => true)
    if (!assignment) return {}

    const componentOptions = getComponentOptions()
    const componentName = getComponentName()

    let props: Prop[] = []
    let emits: Emit[] = []
    if (isDefineComponentCall()) {
      const type = callExpr!.getType()
      const [propsType, , , , , , , emitsType] = type.getAliasTypeArguments()
      props = analyzePropsType(propsType)
      emits = analyzeEmitsType(emitsType)
    }

    return {
      componentName,
      props,
      emits,
    }
  }

  function resolvePropRawType(file: SourceFile, type: Type) {
    const resolvedType = file.addTypeAlias({
      name: '__RESOLVE_PROPS_RAW_TYPE',
      type: `ResolvePropType<${type.getText()}>`,
    })
    const resolved = resolveType(resolvedType.getType())
    resolvedType.remove()
    return resolved
  }

  function resolveType(type: Type): ResolvedType {
    let refs: TypeRef[] | undefined = undefined
    const decls = type.getSymbol()?.getDeclarations()
    refs = (decls ?? []).map((decl) => {
      const start = decl.getStart()
      const end = decl.getEnd()
      let file = decl.getSourceFile().getFilePath().toString()
      if (
        rootPath &&
        !file.startsWith('/node_modules') &&
        path.isAbsolute(file)
      ) {
        file = path.relative(rootPath, file)
      }
      const text = !isTsLib(file) ? decl.getText() : undefined
      return {
        start,
        end,
        file,
        text,
      }
    })

    let unionType: ResolvedType[] | undefined = undefined
    if (type.isUnion()) {
      unionType = type.getUnionTypes().map((type) => resolveType(type))
    }
    return {
      refs,
      unionType,
      text: type.getText(
        undefined,
        TypeFormatFlags.InTypeAlias | TypeFormatFlags.NoTruncation
      ),
    }
  }
}
