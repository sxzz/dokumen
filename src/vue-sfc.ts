import { parse } from '@vue/compiler-sfc'
import { parseTS } from './ts'
import type { Options } from '.'
import type { SourceFile } from 'ts-morph'

export const parseVueSFC = (sourceFile: SourceFile, options: Options) => {
  const source = sourceFile.getText()
  const { descriptor } = parse(source, {
    filename: sourceFile.getFilePath(),
  })
  let contents = ''
  if (descriptor.script) {
    contents += descriptor.script.content
  }
  if (descriptor.scriptSetup) {
    contents += `\n${descriptor.scriptSetup.content}`
  }
  const project = sourceFile.getProject()
  const file = project.createSourceFile(
    `${sourceFile.getFilePath()}.ts`,
    contents
  )
  project.removeSourceFile(sourceFile)

  return parseTS(file, options)
}
