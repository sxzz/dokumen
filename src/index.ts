import { Project } from 'ts-morph'
import { parseTS } from './ts'
import { parseVueSFC } from './vue-sfc'
import type { Result } from './ts'

export interface Options {
  tsConfigFilePath?: string
  rootPath?: string
}

export const generate = (files: string[], options: Options = {}) => {
  const project = new Project({
    tsConfigFilePath: options.tsConfigFilePath,
    skipAddingFilesFromTsConfig: true,
  })
  project.addSourceFilesAtPaths(files)

  const results: { filePath: string; result: Result }[] = []
  for (const sourceFile of project.getSourceFiles()) {
    const filePath = sourceFile.getFilePath()

    let result: Result
    if (sourceFile.getFilePath().endsWith('.ts')) {
      result = parseTS(sourceFile, options)
    } else if (sourceFile.getFilePath().endsWith('.vue')) {
      result = parseVueSFC(sourceFile, options)
    } else {
      continue
    }

    results.push({
      filePath,
      result,
    })
  }

  return results
}
