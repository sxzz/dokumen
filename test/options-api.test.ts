import path from 'path'
import { expect, it } from 'vitest'
import { generate } from '../src'

const rootPath = path.join(__dirname, '..')
const tsConfigFilePath = path.join(rootPath, 'tsconfig.json')
const pathFixtures = path.resolve(__dirname, 'fixtures')

it('Options API', () => {
  const results = generate([path.resolve(pathFixtures, 'options.vue')], {
    tsConfigFilePath,
    rootPath,
  }).map(({ result }) => result)
  expect(results).matchSnapshot()
})
