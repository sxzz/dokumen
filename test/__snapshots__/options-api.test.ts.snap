// Vitest Snapshot v1

exports[`Options API 1`] = `
[
  {
    "componentName": "ElHello",
    "emits": [
      {
        "description": "onclick",
        "name": "click",
        "signatures": [
          [
            {
              "description": "事件",
              "name": "evt",
              "type": {
                "refs": [
                  {
                    "end": 227286,
                    "file": "/node_modules/typescript/lib/lib.dom.d.ts",
                    "start": 226435,
                    "text": undefined,
                  },
                  {
                    "end": 227390,
                    "file": "/node_modules/typescript/lib/lib.dom.d.ts",
                    "start": 227298,
                    "text": undefined,
                  },
                ],
                "text": "MouseEvent",
                "unionType": undefined,
              },
            },
            {
              "description": "名称",
              "name": "name",
              "type": {
                "refs": [],
                "text": "string",
                "unionType": undefined,
              },
            },
          ],
        ],
      },
      {
        "description": "123",
        "name": "remove",
        "signatures": [
          [
            {
              "description": "姓名",
              "name": "name",
              "type": {
                "refs": [],
                "text": "string",
                "unionType": undefined,
              },
            },
          ],
          [
            {
              "description": "",
              "name": "key",
              "type": {
                "refs": [],
                "text": "number",
                "unionType": undefined,
              },
            },
          ],
        ],
      },
    ],
    "props": [
      {
        "default": "",
        "description": "foo 描述",
        "name": "foo",
        "required": true,
        "type": {
          "refs": [],
          "text": "string",
          "unionType": undefined,
        },
      },
      {
        "default": "123",
        "description": "",
        "name": "bar",
        "required": false,
        "type": {
          "refs": [],
          "text": "number",
          "unionType": undefined,
        },
      },
      {
        "default": "\\"123\\"",
        "description": "测试 baz
测试123",
        "name": "baz",
        "required": false,
        "type": {
          "refs": [],
          "text": "string",
          "unionType": undefined,
        },
      },
      {
        "default": "",
        "description": "",
        "name": "custom",
        "required": true,
        "type": {
          "refs": [
            {
              "end": 792,
              "file": "test/fixtures/options.vue.ts",
              "start": 766,
              "text": "() => boolean | CustomType",
            },
          ],
          "text": "() => boolean | CustomType",
          "unionType": undefined,
        },
      },
      {
        "default": "",
        "description": "",
        "name": "date",
        "required": false,
        "type": {
          "refs": [
            {
              "end": 8274,
              "file": "/node_modules/typescript/lib/lib.es5.d.ts",
              "start": 6960,
              "text": undefined,
            },
            {
              "end": 8656,
              "file": "/node_modules/typescript/lib/lib.es5.d.ts",
              "start": 8636,
              "text": undefined,
            },
            {
              "end": 46260,
              "file": "/node_modules/typescript/lib/lib.es5.d.ts",
              "start": 45984,
              "text": undefined,
            },
            {
              "end": 880,
              "file": "/node_modules/typescript/lib/lib.es2015.symbol.wellknown.d.ts",
              "start": 686,
              "text": undefined,
            },
          ],
          "text": "Date",
          "unionType": undefined,
        },
      },
      {
        "default": "",
        "description": "",
        "name": "mixed",
        "required": false,
        "type": {
          "refs": [
            {
              "end": 54,
              "file": "test/fixtures/_utils/types.ts",
              "start": 0,
              "text": "export interface Interface {
  a: string
  b: number
}",
            },
          ],
          "text": "Interface",
          "unionType": undefined,
        },
      },
    ],
  },
]
`;
